Adding Alan's  sequence kernels to eigenpro

OshanJayawardena opened the pull request:
The end goal is adding Alan's sequence kernels to eigenpro

I will try to do some refactoring at the same time

OshanJayawardena committed:
working stage

OshanJayawardena committed:
sequence kernels added

s1van commented on examples/pe_example.py:
```diff
@@ -0,0 +1,47 @@
+    x = x.cpu().numpy()
+    z = z.cpu().numpy()
+    return kernels.hamming_ker_imq(x, seqs_y=z, alphabet_name='prot', scale=1, beta=1/2, lag=1)
+    # return kernels.laplacian(x, z, bandwidth=20.0)
```
can we remove this unused line

OshanJayawardena replied:
Yes, cleaned up the code

s1van commented on examples/pe_example.py:
what does `pe` stand for in `pe_example`?
do you think we can rename it to `sequence_kernel_example`?

OshanJayawardena replied:
yes, I renamed it

s1van commented on eigenpro/utils/eigh.py:
```diff
@@ -97,7 +97,7 @@ def top_q_eig(matrix: torch.Tensor, q: int) -> EigenSystem:
     n_sample = matrix.shape[0]
     eigenvalues, eigenvectors = sp.linalg.eigh(
-        matrix, eigvals=(n_sample - q-1, n_sample - 1))
+        matrix, subset_by_index=(n_sample - q-1, n_sample - 1))
```
Thanks for finding this argument change. We really need to add some unittest soon to capture it.

OshanJayawardena replied:
I added unit tests for all kernels.
Also moved the tests for eigenpro.data to test folder
also added a new tests for a new data.util method

s1van commented on eigenpro/kernels/utils.py:
```diff
@@ -0,0 +1,350 @@
+################ Hamming distances #################
+
+
+def hamming_dist_slow(seqs_x, seqs_y, alphabet_name='dna', lag=1):
```
can we remove these two blank lines?

OshanJayawardena replied:
I ruff formatted the entire repo, now all spaces should be fixed.
I also added a ruff config to vscode settings, now vscode will automatically format codes on save using ruff formatter

s1van commented on eigenpro/kernels/utils.py:
```diff
@@ -0,0 +1,350 @@
+import torch
+
+
+alphabets_en = {
```
Could you add a reference link in comment (to articulate the choice of the symbols)?

OshanJayawardena replied:
I moved this to eigenpro.data.utils and I added the link there

s1van commented on eigenpro/kernels/utils.py:
```diff
@@ -0,0 +1,350 @@
+
+################ Reformat axes of sequence representations #############
+
+
```
can we remove these two blank lines?

OshanJayawardena replied:
Sorry I got rid of the entire file, because now I'm using a pytorch implementation of imq

s1van commented on examples/pe_example.py:
```diff
@@ -0,0 +1,47 @@
+import eigenpro.utils.device as dev
+from datasets import load_dataset
+
+n, p, d, c = 1000, 100, 5, 2
```
what does n, p, d, c stand for?
could you add a comment to explain or include a reference?

OshanJayawardena replied:
I added these

s1van reviewed (commented):
I will try running the code today. If it works, I will approve the changes.

OshanJayawardena commented:
Thank you very much for the review, this is still a work in progress, I need to fix few things in sequence kernels yet.

I will fix these and address your review at the same time.

I'm happy to add some unit tests in a seperate PR. We can divide modules and write tests. I can start with unit tests for kernels, because I'm familiar with it. Let me know what you think

s1van commented:
Hi Oshan, it would be great if you can add some unit tests.
Yes, having a separate PR for only sequence kernels and its unit tests is a great idea.

OshanJayawardena committed:
imq pytorch, tests, ruff formatting

OshanJayawardena commented:
I did a lot of changes:

- Adapted a pytorch version of sequence kernel from Gpytorch repo
- So, got rid of all numpy implementations
- Ruff formatted the entire repo
- Added a ruff configuration to .vscode/settings.json, now all codes are formatted on save by vscode
- Added a unittest configuration to .vscode/settings.json, now we can use vscode testing and debug tools
- Added unit tests for all kernels, right now they don't asset expected values. They only check if kernels run with different input combinations
- Moved tests for data that was in eigenpro/data to test/unit/test_data

s1van commented on .vscode/settings.json:
Could you remove this file from the pulling request for now? This is more of a local setting file for dev.

s1van commented on eigenpro/data/__init__.py:
I don't think this file is needed. can we remove it?

s1van commented on eigenpro/data/array_dataset.py:
For refactoring, I would prefer using a separate pulling request. Can we let this PR include only changes related to the sequence kernel?

s1van commented on eigenpro/models/kernel_machine.py:
This is only related to refactoring. Please do the code refactoring in a separate PR.

s1van commented on eigenpro/models/preallocated_kernel_machine.py:
same here. please remove this file from the current PR.

s1van commented on eigenpro/models/sharded_kernel_machine.py:
Please do the code refactoring in a separate PR.

s1van commented on eigenpro/optimizers.py:
Please do the code refactoring in a separate PR.

s1van commented on eigenpro/projector.py:
same here.

s1van commented on eigenpro/solver.py:
same here.

s1van commented on eigenpro/solver.py:
```diff
@@ -14,11 +14,25 @@
-        tmp_centers_coeff=2, wandb=None, T=None, epochs=1,
-        accumulated_gradients=True):
+def fit(
+    model,
```
We typically format the code based on Google Python Style. It is not required to put each argument in a single line.

s1van commented on eigenpro/utils/cache.py:
same here.

s1van commented on eigenpro/utils/device.py:
same here

s1van commented on eigenpro/utils/eigh.py:
same here

s1van commented on eigenpro/utils/fmm.py:
same here.

s1van commented on eigenpro/utils/keigh.py:
Please do refactoring in a separate PR.

s1van commented on eigenpro/utils/mapreduce.py:
Please do refactoring in a separate PR.

s1van commented on eigenpro/utils/metrics.py:
Please do refactoring in a separate PR.

s1van commented on eigenpro/utils/ops.py:
Please do refactoring in a separate PR.

s1van commented on examples/notebook.ipynb:
Do you need to check in your local run result? I think there is no need to change this file.

s1van commented on examples/run_example.py:
Please do refactoring in a separate PR.

s1van commented on examples/run_example.py:
```diff
@@ -14,28 +14,61 @@
 
 def parse_cmd_args():
     parser = argparse.ArgumentParser()
-    parser.add_argument("--n_train", type=int, help="Number of train samples", default=50_000)
```
I prefer the more compact presentation.

s1van commented on test/unit/test_kernels.py:
Thanks for adding the unit tests. Can we name the unit test file as follows:
For your sequence kernel, name the file
kernels/test_standard_kernels.py

You don't need to add the unit tests for the existing code (while it is helpful if you have the bandwidth).
In addition, the current unit tests only check if the functions can return. Could you also check the return value (so it won't change unexpectedly)? e.g.,

@pytest.mark.parametrize(
    "vocab_size, batch_shape",
    [(3, torch.Size([1])), (2, torch.Size([2])), expected],
)
def test_hamming_imq(x_1hot, z_1hot, vocab_size, batch_shape, expected):
    assert_equal(hamming_imq(x_1hot, z_1hot, vocab_size, batch_shape),
                         expected)

s1van commented on examples/sequence_kernel_example.py:
```diff
@@ -0,0 +1,69 @@
+import eigenpro.solver as solver
+import eigenpro.utils.device as dev
+from eigenpro.data.utils import protein_2_1hot
+from datasets import load_dataset
```
what is the `datasets` package? is this huggingface?

OshanJayawardena replied:
yes it is hugging face. quite handy

s1van reviewed (commented):
Sorry for the late reply. Here are a few comments:
1. Could you remove some files for refactoring from this PR (see detailed comments)? 
2. Do you want to add a separate notebook under examples for the sequence kernel?

It is preferred changing as few files as possible in one PR so it is easier for review and approval.
